coreline_ftco_pos_ex2/static/src/models.js

```diff
@@ -0,0 +1,81 @@
+/** @odoo-module **/
+import models from "point_of_sale.models";
+
+models.load_models([{
+    model:  'pos.bundle',
+    fields: ['name', 'product_ids'],
+    domain: [['product_ids', '!=', false]],
+    loaded: function (self, bundles) {
+        self.bundles = bundles;
+        if (!self.config.limit_categories) {
+            self.bundles = bundles;
+        } else {
+            self.bundles = [];
+            const pos_product_ids = Object.keys(self.db.product_by_id).map(key => parseInt(key));
+            for (let bundle of bundles) {
+                let result = bundle.product_ids.every(function(product) {
+                    return pos_product_ids.includes(product);
+                });
+                if (result){
+                    self.bundles.push(bundle);
+                }
+            }
+        }
+    },
+}]);
```

Comment:
This could be simplified a little bit further. 

1. `load_models()` takes a 2<sup>nd</sup> optional arg that allows you to control the loading order of models. You could ensure that you have access to the product ids when your `domain` and `loaded` functions fire
2. you could reduce the data you fetch from the backend by passing the product ids to the domain
3. you could use `Set` to make your flow have a smaller time complexity.
    -  `Array.prototype.includes()` x `Array.prototype.every()` x `for loop over bundles` results into a time complexity of O(n<sup>3</sup>)
    - If you place your allowed ids into a some sort of a hash map with search functionality with complexity of O(1), that would replace your `Array.prototype.includes()` you'd end up with O(n<sup>2</sup>)
4. taking a functional approach over a procedural approach would always make your code much easier to maintain since the next developer will be able to know what you are doing by simply reading the functions name => replace the for loop with an `Array.prototype.filter()` call

```suggestion
models.load_models([{
    model:  'pos.bundle',
    fields: ['name', 'product_ids'],
    domain: function(self) {
        return [
            ["product_ids", "!=", False],
            ["product_ids", "in", Object.keys(self.db.product_by_id)]
        ]      
    },
    loaded: function (self, bundles) {
        const allowedProductIds = new Set(Object.keys(self.db.product_by_id).map(parseInt));
        self.bundles = bundles.filtered(bundle => bundle.product_ids.every(productId => allowedProductIds.has(productId)));
    },
}]);
```